lib/Dialect/FIRRTL/FIRRTLConnectionGraph.cpp

```diff
@@ -0,0 +1,222 @@
+  while (resultIterator == resultIteratorEnd && resultIndex != resultIndexEnd) {
+    ++resultIndex;
+    if (resultIndex == resultIndexEnd)
+      break;
+    auto result = op->getResult(resultIndex);
+    resultIterator = result.use_begin();
+    resultIteratorEnd = result.use_end();
+    while (resultIterator != resultIteratorEnd &&
+           dyn_cast<FConnectLike>(resultIterator.getUser()) &&
+           resultIterator.getOperand()->getOperandNumber() == 0)
+      ++resultIterator;
+  }
+}
+
+bool ResultIterator::operator==(const ResultIterator &other) const {
+  if (resultIndex == resultIndexEnd)
+    return other.resultIndex == other.resultIndexEnd;
+
+  return resultIterator == other.resultIterator &&
+         resultIteratorEnd == other.resultIteratorEnd &&
+         resultIndex == other.resultIndex &&
+         resultIndexEnd == other.resultIndexEnd;
+}
+
+FIRRTLOperation &ResultIterator::operator*() const {
+  assert(resultIndex != resultIndexEnd && "tried to dereference past the end");
+  return *resultIterator->getOwner();
+}
+
+ResultIterator &ResultIterator::operator++() {
+  assert(resultIndex != resultIndexEnd &&
+         "incrementing past the end of the iterator");
+  ++resultIterator;
+  fastforward();
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+//
+// ConnectionIterator
+//
+//===----------------------------------------------------------------------===//
+
+ConnectionIterator::ConnectionIterator(FIRRTLOperation *op, bool empty) {
+  assert(op && "op should not be null");
+  TypeSwitch<FIRRTLOperation *>(op)
+      .Case<FModuleOp>([&](auto op) {
+        iterator = empty ? FModuleOpIterator() : FModuleOpIterator(op);
+      })
+      .Case<FConnectLike>([&](auto op) {
+        iterator = empty ? FConnectLikeIterator() : FConnectLikeIterator(op);
+      })
+      .Default([&](auto *op) {
+        iterator = empty ? ResultIterator() : ResultIterator(op);
+      });
+  this->op = op;
+}
+
+bool ConnectionIterator::operator==(const ConnectionIterator &other) const {
+  assert(iterator.index() == other.iterator.index() &&
+         "comparing different iterator variants");
+  switch (iterator.index()) {
+  case 0:
+    assert(false && "should be impossible");
+    break;
+  case 1:
+    return std::get<FModuleOpIterator>(iterator) ==
+           std::get<FModuleOpIterator>(other.iterator);
+  case 2:
+    return std::get<FConnectLikeIterator>(iterator) ==
+           std::get<FConnectLikeIterator>(other.iterator);
+  case 3:
+    return std::get<ResultIterator>(iterator) ==
+           std::get<ResultIterator>(other.iterator);
+  }
+  return false;
+}
+
+bool ConnectionIterator::operator!=(const ConnectionIterator &other) const {
+  return !(*this == other);
+}
+
+FIRRTLOperation *ConnectionIterator::operator*() {
+  FIRRTLOperation *result;
+  switch (iterator.index()) {
+  case 0:
+    assert(false && "should be impossible");
+    break;
+  case 1:
```

Comment:
I don't know best practices for `std::variant`, but is there a way to not bake in the integer indexes of the variant kinds here?